src/testRunner/unittests/services/extract/ranges.ts

```diff
@@ -14,85 +14,100 @@ namespace ts {
-        const expectedRange = t.ranges.get("extracted");
-        if (expectedRange) {
-            let pos: number, end: number;
-            const targetRange = result.targetRange!;
-            if (isArray(targetRange.range)) {
-                pos = targetRange.range[0].getStart(f);
-                end = last(targetRange.range).getEnd();
+    function testExtractRange(caption: string, s: string) {
+        return it(caption, () => {
+            const t = extractTest(s);
+            const f = createSourceFile("a.ts", t.source, ScriptTarget.Latest, /*setParentNodes*/ true);
+            const selectionRange = t.ranges.get("selection");
+            if (!selectionRange) {
+                throw new Error(`Test ${s} does not specify selection range`);
+            }
+            const result = refactor.extractSymbol.getRangeToExtract(f, createTextSpanFromRange(selectionRange));
+            const expectedRange = t.ranges.get("extracted");
+            if (expectedRange) {
+                let pos: number, end: number;
+                const targetRange = result.targetRange!;
+                if (isArray(targetRange.range)) {
+                    pos = targetRange.range[0].getStart(f);
+                    end = last(targetRange.range).getEnd();
+                }
+                else {
+                    pos = targetRange.range.getStart(f);
+                    end = targetRange.range.getEnd();
+                }
+                assert.equal(pos, expectedRange.pos, "incorrect pos of range");
+                assert.equal(end, expectedRange.end, "incorrect end of range");
             }
             else {
-                pos = targetRange.range.getStart(f);
-                end = targetRange.range.getEnd();
+                assert.isTrue(!result.targetRange, `expected range to extract to be undefined`);
             }
-            assert.equal(pos, expectedRange.pos, "incorrect pos of range");
-            assert.equal(end, expectedRange.end, "incorrect end of range");
-        }
-        else {
-            assert.isTrue(!result.targetRange, `expected range to extract to be undefined`);
-        }
+        });
     }
 
     describe("unittests:: services:: extract:: extractRanges", () => {
-        it("get extract range from selection", () => {
-            testExtractRange(`
+        describe("get extract range from selection", () => {
+            testExtractRange("extractRange1", `
                 [#|
                 [$|var x = 1;
                 var y = 2;|]|]
             `);
-            testExtractRange(`
-                [#|
-                var x = 1;
-                var y = 2|];
+            testExtractRange("extractRange2", `
+                [$|[#|var x = 1;
+                var y = 2|];|]
             `);
-            testExtractRange(`
-                [#|var x = 1|];
+            testExtractRange("extractRange3", `
+                [#|var x = [$|1|]|];

```

Comment:
This test makes me think `[$|` is the selection range and `[#|` is the extraction range, but the next test makes me think the opposite.  Which is which?

Reply:
Oh, maybe it's the opposite and this test actually _shrinks_ the selection range?  How does this interact with your change to expand to include the semicolon?

Reply:
`[#|` is the selection range and `[$|` is the extraction range. The extraction range should *always* have shrunk to only the initializer because that is the range that is actually extracted. This should have been a valid test before, but it didn't exist:

`[#|var x = [$|1|];|]`

In this case, the entire variable statement is selected which simply gets checked for an initializer.

The test changed here was meant to show that the refactor does not work if the semicolon is not selected (because the entire statement had to be within the span). My semicolon change simply ignores this weird case and pretends you selected the semicolon.

Reply:
Actually, the test *does* exist, see `extractRange22`.